Add support for featurecollections and expressions in geojson format

fhissink opened the pull request:
This PR contains changes in:
- addSource
- addLayer
- handling click events on the map
- queryRenderedFeatures

By changing `addSource` to support FeatureCollections the `circles` array which kept track of the added source was no longer useable. Therefor some changes needed to be made to the way click events on the map are handled.
As an added bonus this also solved the issue where line features needed to have a seperate way to handle click events (with the annotationlayer). So this seems no longer needed, although I don't know if there are any edge cases that I might miss now.
Once you click on the map all layers that are registered to the mapEvent
`this.mapboxView.onMapEvent( 'click', 'layer-id', ...` will be queried, this means that if multiple layers are on top of each other, multiple events could be triggered. 
I've added a sample of this in the demo-angular project (where a line, point and featurecollection layer are on the of each other).

Since the click events on the map don't support async functions I needed to convert `queryRenderedFeatures` from an async/promise function to a 'sync' function. Since this code was already synchronous from mapbox itself this wasn't a very big change. Also, since only the features that are already rendered on the map are being queried I don't think having this function synchronous would cause performance issues.

CircleLayers now support expressions in json format, all the paint property that were already supported now also support json expressions.
The same could probably also work for linelayers but that hasn't been implemented yet.

Still left to do: 
- addLinePoint -> this is still commented out because this relied on the lines array which is no longer available -> **Done**
- json expression support for linelayers
- I don't own any Apple devices so I'm not in the position to develop/debug/test on iOS, so that part hasn't changed

N.B.: Sorry for the whitespace changes you check it without the whitespaces by adding `?w=1` to the url

lis-bv committed:
Add support for featurecollections and expressions in geojson format

fhissink commented on src/mapbox.android.ts:
```diff
@@ -3440,62 +3162,63 @@ export class Mapbox extends MapboxCommon implements MapboxApi {
   */
 
-  public addLinePoint( id : string, lnglat, nativeMapView? ) : Promise<any> {
+  public addLinePoint(id: string, lnglat, nativeMapView?): Promise<any> {
```
This function is commented out and needs some work still to make it work again

Yermo commented:
Oh cool. Thank you for this. Hopefully I'll have a chance to look at this in a bit more detail tomorrow.

Yermo commented:
We'll need to make sure this is synced up with iOS side.

fhissink commented:
> We'll need to make sure this is synced up with iOS side.

Yes, would be nice if I could get some help with that. I'll try to fix the other remaining tasks somewhere this week.

Yermo commented:
@fhissink absolutely. If I remember you don't have access to iOS development tools? I was pretty much expecting to have to do the iOS side. The expression support you've come up with should, hopefully, dramatically simplify things and, as you mention, allow us to remove my gawdawful annotation hack for click events. Did you test the click support when zooming in and out?

fhissink commented:
> @fhissink absolutely. If I remember you don't have access to iOS development tools? I was pretty much expecting to have to do the iOS side. The expression support you've come up with should, hopefully, dramatically simplify things and, as you mention, allow us to remove my gawdawful annotation hack for click events. Did you test the click support when zooming in and out?

@Yermo, thanks for picking that up:). Regarding the testing when zooming in and out, I haven't tested it explicitly yet, but I definitely did zoom before clicking (I assume you mean **after** a zoom action happened and not **while**?). I've got some time to work on this tomorrow again so I can also give you a better answer to this. Are there any other scenario's I need to look at?

Yermo commented:
@fhissink It's been a while since I've written that code, but the whole reason I implemented that second layer kludge was that I couldn't figure out how to have the click area zoom with the map. My suspicion is that your fix addresses that. So basically in a test add a click handler to a circle and to a line and make sure the click behavior works as you would expect it to as you zoom in and out. If it does we should be able to just get rid of that horrible double layer kludge of mine.

fhissink commented:
Ok, I think that all works, but I'll double check it tomorrow.

lis-bv committed:
Add implementation for addLinePoint

fhissink commented:
@Yermo, I've tested the click behaviour and that all seems to work fine.

I've just added the implementation for the addLinePoint function. It's still a bit hacky mainly because of the limitations of `querySourceFeatures` (it only returns already rendered features).

The function itself was (and still is) a bit limited. It only works for LineLayers that have only 1 feature.
I changed the logic a bit because in the function itself it was searching for layer with id `x` and source with id `x_source`, I'm not sure what the exact reason for that is, but I've changed it a bit so that you have to provide the id of the source directly (the getting of the layer will be skipped), that way the naming of the source layer is a bit less strict (`_source` doesn't need to be in the id anymore). I can change it back if you prefer that, but I didnt see the benefits of it.

fhissink commented:
@Yermo, did you manage to have a look at the code already? I was wondering what you think of it, and if you think you can apply the same logic on the iOS side? Regarding that do you have any idea when you will be able to work on that part?

Another question; I would also like to add another feature, which is the posibility to update a source (which would make the data update 'real-time'). I was wondering if I should add it to this PR as well, or if we should first finish the stuff there is now and add new functionality later?

Yermo commented:
@fhissink I apologize. I've taken a quite look at it but have been too head deep in my project to focus on it. I hope to find some time to look at it in more detail in the next few days. On cursory examination it seems solid to me. I'm wondering if there's a way to parse the entire style specification in one call instead of just parsing the expressions. If we could do this then it would go a very long way to making the native API match the Web GL JS api. 

Let's finish this pull request up first and then we can focus on updating sources.

Yermo commented:
So I finally have some time to look at this. Unfortunately, because of the formatting changes the diff doesn't clearly show the changes that you made. I was wondering if you could re-apply your changes so that it's easier to see what was changed?

fhissink commented:
Yes, I'll try that. I have some time for that tomorrow.

fhissink commented:
Hi @Yermo, unfortunately I didn't get to re-applying the changes. If you add ?w=1 as a querystring parameter in the url you can see all the changes without the whitespaces. Can you maybe work with that already?
[https://github.com/Yermo/nativescript-mapbox/pull/359/commits/0e1a83ff1cee194c50f8757bb1c4f4f9b2f18b3b?w=1 ](https://github.com/Yermo/nativescript-mapbox/pull/359/commits/0e1a83ff1cee194c50f8757bb1c4f4f9b2f18b3b?w=1)

fhissink commented:
@Yermo, I created a new PR without the whitespace changes
